src/service/viewport/viewport-impl.js

```diff
@@ -396,42 +401,50 @@ export class Viewport {
   }
 
   /**
-   * Returns a promise that resolves with latest rect of the element within the document.
+   * Returns the client rect of the element
+   * TODO(@zhouyx): Consider combine with #getClientRectInViewportAsync()
```

Comment:
I think we need to address this `TODO` right away. One algo for now should do. We can parametrize with arguments want we want to do with intersection edge cases. E.g. if an element is outside of local client rect, should we just consider it outside of global client rect as well? E.g. when `local.top < 0`, is there a really good reason to add it to `global.top`? What does that number mean in that case?

Reply:
I do agree the `clientRect` when element is outside root is meaningless and incorrect.
So I combine these two methods, kind of... 

- always calculate the the root clientRect and the element boundingClientRect. 
- If root doesn't exist (ampdoc case), always return the element's boundingClientRect.
- If the root does exist (inabox case). first check if the root intersects with the element. And return element clientRect to global viewport or null based on the result. 

Still I don't know if this is good. It is kind of inconsistent between ampdoc and inabox. I think the ideal method should always return null if element is outside global viewport otherwise return its clientRect. But I can't go ahead and implement this right away, because for inabox, we always expect host to send its parent iframe's clientRect even when it is outside viewport. 😢. 

Another concern is that `getClientRectAsync()` sounds like a method that should always return to me a `clientRect`. Which is not the case here. This is probably OK since `clientRect` can be pretty meaningless when element doesn't intersect with viewport.